lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+	 *         'id' => array( 6, 17, 'test-4' ),
+	 *         'class' => array( 18, 32, 'outline' )
+	 *     );
+	 * </code>
+	 *
+	 * @var array<WP_HTML_Attribute_Token>
+	 */
+	public $attributes = array();
+
+
+	public function __construct( $input ) {
+		$this->document = $input;
+	}
+
+
+	public function scan( WP_Tag_Find_Descriptor $descriptor, $found_already = 0, $tag = null ) {
+		if ( $found_already === $descriptor->match_offset ) {
+			return $tag;
+		}
+
+		$tag = $this->find_next_tag();
+		if ( ! $tag ) {
+			return false;
+		}
+
+		// @TODO: Are we done matching?
+		while ( $attribute = $this->find_next_attribute() ) {
+			// HTML5 says duplicate values are ignored
+			if ( ! array_key_exists( $attribute->name->comparable, $this->attributes ) ) {
+				$this->attributes[ $attribute->name->comparable ] = $attribute;
+			}
+		}
+
+		$does_match = 'matches' === $descriptor->check( $tag->comparable, $this->attributes );
+		$found_this_time = $does_match ? 1 : 0;
+
+		return $this->scan( $descriptor, $found_already + $found_this_time, $tag );
+	}
+
+	/**
+	 * @return WP_HTML_Scanner_Token|false
+	 */
+	public function find_next_tag() {
+		// @TODO: Handle <DOCTYPE>
+		if ( 1 !== preg_match(
+			/*
+			 * Unfortunately we can't try to search for only the tag name we want because that might
+			 * lead us to skip over other tags and lose track of our place. So we need to search for
+			 * _every_ tag and then check after we find one if it's the one we are looking for.
+			 */
+	"~<!--(?>.*?-->)|<!\[CDATA\[(?>.*?>)|<\?(?>.*?)>|<(?P<TAG>[a-z][^\t\x{0A}\x{0C} /?>]*)~Smui",
+			$this->document,
+			$tag_match,
+			PREG_OFFSET_CAPTURE,
+			$this->start_at
+		) ) {
+			return false;
+		}
+
+		list( $full_match, $start_at ) = $tag_match[0];
+
+		// Keep scanning if we found a comment or CDATA section.
+		if ( ! isset( $tag_match['TAG'] ) ) {
+			$this->start_at = $start_at + strlen( $full_match );
+			return $this->find_next_tag();
+		}
+
+		$this->start_at = $start_at + strlen( $full_match );
+		$this->tag_name = WP_HTML_Scanner_Token::from_preg_match( $tag_match['TAG'] );
+		$this->tag_start = $this->tag_name->start - 1; // rewind past the leading `<`
+		$this->attributes = array();
+
+		return $this->tag_name;
+	}
+
+
+	public function find_next_attribute() {
+		// Find the attribute name
+		if ( 1 !== preg_match(
+			'~[\t\x{0a}\x{0c}\x{0d} ]*(?P<NAME>=?[^=/>\t\x{0C} ]*)~Smiu',
+			$this->document,
+			$attribute_match,
+			PREG_OFFSET_CAPTURE,
+			$this->start_at
+		) ) {
+			return false;
+		}
+
+		list( $full_match, $start_at ) = $attribute_match[0];
+
+		if ( empty( $full_match ) ) {
+			return false;
+		}
+
+		$name_token = WP_HTML_Scanner_Token::from_preg_match( $attribute_match['NAME'] );
+
+		$this->start_at = $start_at + strlen( $full_match );
+		if ( '=' !== $this->document[ $this->start_at ] ) {
+			return new WP_HTML_Attribute_Token( $name_token, null, $name_token->start, $start_at + strlen( $full_match ) );
```

Comment:
Similar nitpick as above – it would be nice to have an `end` property:

```suggestion
					return new WP_HTML_Attribute_Token( $name_token, null, $name_token->start, $name_token->end );
```